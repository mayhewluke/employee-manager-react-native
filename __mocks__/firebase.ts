const firebase = jest.genMockFromModule("firebase");

export default firebase;
