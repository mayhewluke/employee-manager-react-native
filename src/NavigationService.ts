import {
  NavigationActions,
  NavigationContainerComponent,
  NavigationParams,
} from "react-navigation";

export type NavigateFunction = (
  routeName: string,
  params?: NavigationParams,
) => boolean;

let rootNavigatorComponent: NavigationContainerComponent;

export function setNavigation(
  navigatorComponent: NavigationContainerComponent,
): void {
  rootNavigatorComponent = navigatorComponent;
}

export const navigate: NavigateFunction = (routeName, params) =>
  rootNavigatorComponent
    ? rootNavigatorComponent.dispatch(
        NavigationActions.navigate({
          params,
          routeName,
        }),
      )
    : false;
