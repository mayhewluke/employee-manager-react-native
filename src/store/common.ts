export type Async<T> =
  | { state: "INIT" }
  | { state: "PROGRESS" }
  | { state: "COMPLETE"; value: T }
  | { state: "ERROR"; error: string };
