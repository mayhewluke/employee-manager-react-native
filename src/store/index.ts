import { applyMiddleware, combineReducers, createStore, Store } from "redux";
import { combineEpics, createEpicMiddleware } from "redux-observable";
import reduxThunk from "redux-thunk";

import { AuthAction, authReducer, IAuthState } from "./Auth";
import {
  createEmployeeEpic,
  employeeReducer,
  fireEmployeeEpic,
  IEmployeeState,
  updateEmployeeEpic,
} from "./Employee";
import {
  employeesReducer,
  employeesSubscriptionEpic,
  IEmployeesState,
} from "./Employees";

const rootReducer = combineReducers({
  auth: authReducer,
  employee: employeeReducer,
  employees: employeesReducer,
});

const rootEpic = combineEpics(
  createEmployeeEpic,
  fireEmployeeEpic,
  updateEmployeeEpic,
  employeesSubscriptionEpic,
);

const epicMiddleware = createEpicMiddleware();

export interface IRootState {
  auth: IAuthState;
  employee: IEmployeeState;
  employees: IEmployeesState;
}

export const store: Store<IRootState, RootAction> = createStore(
  rootReducer,
  {},
  applyMiddleware(reduxThunk, epicMiddleware),
);

epicMiddleware.run(rootEpic);

export type RootAction = AuthAction;
export * from "./common/Async";
