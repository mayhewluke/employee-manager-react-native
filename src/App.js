import React, { Component } from "react";
import { View } from "react-native";
import { Provider } from "react-redux";
import { createStore, applyMiddleware } from "redux";
import ReduxThunk from "redux-thunk";
import firebase from "firebase";

import reducers from "./reducers";
import firebaseJson from "../firebaseConfig.json";

import LoginForm from "./components/LoginForm";

class App extends Component {
  componentDidMount() {
    firebase.initializeApp(firebaseJson);
  }

  render() {
    return (
      <Provider store={createStore(reducers, {}, applyMiddleware(ReduxThunk))}>
        <View>
          <LoginForm />
        </View>
      </Provider>
    );
  }
}

export default App;
