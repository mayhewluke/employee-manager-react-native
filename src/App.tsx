import firebase from "firebase";
import React, { Component } from "react";
import { View } from "react-native";
import { Provider } from "react-redux";

import firebaseConfigJson from "../firebaseConfig.json";
import { store } from "./store";

import LoginForm from "./components/LoginForm";

class App extends Component {
  public componentDidMount() {
    firebase.initializeApp(firebaseConfigJson);
  }

  public render() {
    return (
      <Provider store={store}>
        <View>
          <LoginForm />
        </View>
      </Provider>
    );
  }
}

export default App;
