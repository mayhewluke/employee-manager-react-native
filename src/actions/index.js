import firebase from "firebase";

import {
  EMAIL_CHANGED,
  PASSWORD_CHANGED,
  LOGIN_SUCCESS,
  LOGIN_FAIL,
  LOGIN_START
} from "./types";

export const emailChanged = text => ({
  type: EMAIL_CHANGED,
  payload: text
});

export const passwordChanged = password => ({
  type: PASSWORD_CHANGED,
  payload: password
});

const loginSuccess = dispatch => user =>
  dispatch({ type: LOGIN_SUCCESS, payload: user });

const loginFail = dispatch => () => dispatch({ type: LOGIN_FAIL });

export const logIn = (email, password) => dispatch => {
  dispatch({ type: LOGIN_START });
  firebase
    .auth()
    .signInWithEmailAndPassword(email, password)
    .then(loginSuccess(dispatch))
    .catch(() =>
      firebase
        .auth()
        .createUserWithEmailAndPassword(email, password)
        .then(loginSuccess(dispatch))
        .catch(loginFail(dispatch))
    );
};
