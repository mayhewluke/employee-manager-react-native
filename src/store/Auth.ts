import firebase from "firebase";
import { NavigationScreenProp } from "react-navigation";
import { Dispatch, Reducer } from "redux";
import { action as createAction } from "typesafe-actions";

export enum AuthActionType {
  EMAIL_CHANGED = "EMAIL_CHANGED",
  PASSWORD_CHANGED = "PASSWORD_CHANGED",
  LOGIN_SUCCESS = "LOGIN_SUCCESS",
  LOGIN_FAIL = "LOGIN_FAIL",
  LOGIN_START = "LOGIN_START",
}

export interface IAuthState {
  email: string;
  password: string;
  loading: boolean;
  error: string;
  user: firebase.auth.UserCredential | null;
}

const INITIAL_STATE: IAuthState = {
  email: "",
  error: "",
  loading: false,
  password: "",
  user: null,
};

export type AuthAction =
  | { type: AuthActionType.EMAIL_CHANGED; payload: string }
  | { type: AuthActionType.PASSWORD_CHANGED; payload: string }
  | {
      type: AuthActionType.LOGIN_SUCCESS;
      payload: firebase.auth.UserCredential;
    }
  | { type: AuthActionType.LOGIN_FAIL }
  | { type: AuthActionType.LOGIN_START };

type AuthDispatch = Dispatch<AuthAction>;

export const authReducer: Reducer<IAuthState, AuthAction> = (
  state = INITIAL_STATE,
  action,
) => {
  switch (action.type) {
    case AuthActionType.EMAIL_CHANGED:
      return { ...state, email: action.payload };
    case AuthActionType.PASSWORD_CHANGED:
      return { ...state, password: action.payload };
    case AuthActionType.LOGIN_START:
      return { ...state, loading: true, error: "" };
    case AuthActionType.LOGIN_SUCCESS:
      return { ...state, ...INITIAL_STATE, user: action.payload };
    case AuthActionType.LOGIN_FAIL:
      return { ...state, loading: false, error: "Something went wrong!" };
    default:
      return state;
  }
};

const loginSuccess = (dispatch: AuthDispatch, navigate: any) => (
  user: firebase.auth.UserCredential,
) => {
  dispatch(createAction(AuthActionType.LOGIN_SUCCESS, user));
  navigate("Main");
};

const loginFail = (dispatch: AuthDispatch) => () =>
  dispatch(createAction(AuthActionType.LOGIN_FAIL));

export const emailChanged = (email: string) =>
  createAction(AuthActionType.EMAIL_CHANGED, email);

export const passwordChanged = (password: string) =>
  createAction(AuthActionType.PASSWORD_CHANGED, password);

export const logIn = (
  email: string,
  password: string,
  navigation: NavigationScreenProp<any>,
) => (dispatch: AuthDispatch) => {
  dispatch(createAction(AuthActionType.LOGIN_START));
  firebase
    .auth()
    .signInWithEmailAndPassword(email, password)
    .then(loginSuccess(dispatch, navigation.navigate))
    .catch(() =>
      firebase
        .auth()
        .createUserWithEmailAndPassword(email, password)
        .then(loginSuccess(dispatch, navigation.navigate))
        .catch(loginFail(dispatch)),
    );
};
