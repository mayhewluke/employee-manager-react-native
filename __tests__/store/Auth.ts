import "jest-enzyme";
import React from "react";

import firebase from "firebase";

import { Async } from "../../src/store";
import {
  AuthActionType,
  authReducer as reducer,
  INITIAL_STATE,
} from "../../src/store/Auth";

describe("reducer", () => {
  const testState = {
    email: "user@website.com",
    loginAction: { state: "PROGRESS" } as Async<any>,
    password: "supersecure",
    user: {} as firebase.auth.UserCredential,
  };

  it("sets the initial state", () => {
    // TODO figure out how to test this without violating the type system and
    // without having to create a type specifically just to handle this one time
    // we want to have it do nothing but return the initial state (production
    // code existing just to enable testing is not a good thing)
    expect(reducer(undefined, {})).toEqual(INITIAL_STATE);
  });

  it("updates the email when it is changed", () => {
    const email = "user@website.com";
    const state = reducer(testState, {
      payload: email,
      type: AuthActionType.EMAIL_CHANGED,
    });
    expect(state).toEqual({ ...testState, email });
  });

  it("updates the password when it is changed", () => {
    const password = "supersecure";
    const state = reducer(testState, {
      payload: password,
      type: AuthActionType.PASSWORD_CHANGED,
    });
    expect(state).toEqual({ ...testState, password });
  });

  describe("login flow", () => {
    describe("when login request is started", () => {
      it("sets the asyncAction to in progress", () => {
        const state = reducer(testState, {
          type: AuthActionType.LOGIN_START,
        });
        expect(state).toEqual({
          ...testState,
          loginAction: { state: "PROGRESS" },
        });
      });
    });

    describe("when login is successful", () => {
      it("sets the current user", () => {
        const user = {} as firebase.auth.UserCredential;
        const state = reducer(testState, {
          payload: user,
          type: AuthActionType.LOGIN_SUCCESS,
        });
        expect(state.user).toBe(user);
      });

      it("resets the rest of the state", () => {
        const prevState = {
          email: "foo@bar.com",
          loginAction: { state: "ERROR", error: "" } as Async<any>,
          password: "notsecure",
          user: null,
        };
        const state = reducer(prevState, {
          payload: {} as firebase.auth.UserCredential,
          type: AuthActionType.LOGIN_SUCCESS,
        });
        expect(state.email).toEqual(INITIAL_STATE.email);
        expect(state.password).toEqual(INITIAL_STATE.password);
        expect(state.loginAction).toEqual(INITIAL_STATE.loginAction);
      });
    });

    describe("when login fails", () => {
      it("marks the request as having errored", () => {
        const state = reducer(testState, {
          type: AuthActionType.LOGIN_FAIL,
        });
        expect(state).toEqual({
          ...testState,
          loginAction: { state: "ERROR", error: expect.any(String) },
        });
      });

      it("sets an error message matching a snapshot", () => {
        const state = reducer(testState, {
          type: AuthActionType.LOGIN_FAIL,
        });
        // TODO figure out how to make the type system happy here without
        // confusing verbosity
        expect(state.loginAction.error).toMatchSnapshot();
      });
    });
  });
});
