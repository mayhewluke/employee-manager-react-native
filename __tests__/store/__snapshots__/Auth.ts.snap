// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer login flow when login fails sets an error message matching a snapshot 1`] = `"Something went wrong!"`;
